tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -1362,12 +1370,12 @@ static void resolveCursor(SwiftLangSupport &Lang,
     void cancelled() override {
       CursorInfoData Info;
       Info.IsCancelled = true;
-      Receiver(Info);
+      Receiver(Info, "Cancelled.");
```

Comment:
Here the `Receiver` receives a `"Cancelled."` error string, which would normally translate to an error response, but the `Info.IsCancelled` bit is set to indicate that the response should fail with `RequestCancelled` instead.

I have a suggestion to make the code somewhat more readable and maintainable long term. Instead of having the `Receiver`s accept a data object and a `StringRef` separately, how about we have something like a `Result<T>` object, which can also encapsulate:
* The data object
* or an error message
* or indication that the operation was cancelled

Roughly something like this:
```
template <typename T>
class Result {
  static Result fromData(const T *data) // can point to a data object
  static Result error(StringRef error) // or can wrap an error string
  static Result cancelled() // or can indicate cancellation

  const T &data() const // asserts it has data 
  bool isError() const
  StringRef getError() const
  bool isCancelled() const
<...>
};
```
Then this is what you pass to the receivers, like `Receiver(Result::error("")` or `Receiver(&Info)`

What do you think?

Reply:
Yeah that seems reasonable, so remove the `IsCancelled` field from the infos and instead add it to the `Result` type (or equivalent)? I'll have to double check to see how cleanly this can be done but it sounds good.

Reply:
> so remove the IsCancelled field from the infos and instead add it to the Result type (or equivalent)? 

Yes, exactly.

Reply:
Done, let me know what you think